text/0004-template-groups.md

```diff
@@ -0,0 +1,105 @@
+
+# Detailed design
+
+We update the APIs to add semantics to the `@templategroup` annotation, as shown in the [Basic example](#basic-example). In particular, we will need to update the APIs for linking template policies to consider the `@templategroup` annotation, and not just the policy ID. A linkage for group `role1` should link any template having that group. A linkage for ID `policy1` should _fail_ if `policy1` also has a `@templategroup` annotation -- this is important for forbidding linkages to individual group policies.
+
+The updated APIs are for the CLI and core Cedar.
+
+# Drawbacks
+
+The main reason not to do this is that it adds extra code that customers _could_ implement themselves, using the existing annotation mechanism. However, doing so would lose the benefits of making the mechanism known to UIs, analysis tools, etc. and it would impose extra work on customers.
+
+
+# Alternatives
+
+We could add new syntax to Cedar that supports policy groups. For example
+```
+@id("role1")
+{ 
+    @id("policy1")
+    permit(
+      principal == ?principal,
+      action, 
+      resource in ?resource);
+
+    @id("policy2")
+    permit(
+      principal == ?principal,
+      action == Action::"viewDoc",
+      resource in Directory::"DivisionDocs");
+}
+```
+Here we allow any statements to by syntactically grouped within braces, and add an annotation on the group. This is potentially more visually clear to readers, and evokes [AWS IAM policies, which can consist of multiple statements grouped together](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_statement.html). But it would require extending the Cedar grammar.
+
+We could opt to do nothing at the level of the language, and rely on users of Cedar (such as Amazon Verified Permissions) to implement grouping using metadata of their choice, even as suggested here. We could also add the grouping syntax above, but not more than that, again relying on users to interpret the metadata annotations themselves. However, if grouping policies is likely to be generally useful, it seems useful to provide support for it.
+
+Users can also achieve the desired grouping effect by combining both templates into a single one:
+```
+@id("bothpolicies")
+permit(
+  principal == ?principal,
+  action, 
+  resource)
+when {
+   resource in ?resource ||
+   (action == Action::"viewDoc" && resource in Directory::"DivisionDocs")
+};
+```
+This has the drawback that the combined policy is more difficult to understand. (It also requires the template slot `?resource` to appear in the `when` condition, rather than the scope, which [RFC-0003](https://github.com/cedar-policy/rfcs/blob/main/text/0003-placeholders-in-conditions.md
```

Comment:
Link doesn't work (and won't until such time as RFC 3 gets merged); instead, you could consider using the link https://github.com/cedar-policy/rfcs/pull/3

Reply:
Fixed.